src/Parsers/IAST.cpp

```diff
@@ -28,34 +28,86 @@ const char * IAST::hilite_none         = "\033[0m";
 
 IAST::~IAST()
 {
-    /// If deleter was set, move our children to it.
-    /// Will avoid recursive destruction.
-    if (deleter)
+    /** Create intrusive linked list of children to delete.
+      * Each ASTPtr child contains pointer to next child to delete.
+      */
+    ASTPtr delete_list_head_holder = nullptr;
+    bool delete_directly = next_to_delete_list_head == nullptr;
+    ASTPtr & delete_list_head_reference = next_to_delete_list_head ? *next_to_delete_list_head : delete_list_head_holder;
+
+    /// Move children into intrusive list
+    for (auto & child : children)
     {
-        deleter->push_back(std::move(children));
-        return;
+        /** If two threads remove ASTPtr concurrently,
+          * it is possible that neither thead will see use_count == 1.
+          * It is ok. Will need one more extra stack frame in this case.
+          */
+        if (child.use_count() != 1)
+            continue;
+
+        ASTPtr child_to_delete;
+        child_to_delete.swap(child);
+
+        if (!delete_list_head_reference)
+        {
+            /// Initialize list first time
+            delete_list_head_reference = std::move(child_to_delete);
+            continue;
+        }
+
+        ASTPtr previous_head = std::move(delete_list_head_reference);
+        delete_list_head_reference = std::move(child_to_delete);
+        delete_list_head_reference->next_to_delete = std::move(previous_head);
     }
 
-    std::list<ASTs> queue;
-    queue.push_back(std::move(children));
-    while (!queue.empty())
+    if (!delete_directly)
+        return;
+
+    while (delete_list_head_reference)
     {
-        for (auto & node : queue.front())
+        auto next_node_to_delete = std::move(delete_list_head_reference->next_to_delete);
+
+        /** Pass list head into child before destruction.
+          * It is important to properly handle cases where subclass has member same as one of its children.
+          *
+          * class ASTSubclass : IAST
+          * {
+          *     ASTPtr first_child; /// Same as first child
+          * }
+          *
+          * In such case we must move children into list only in IAST destructor.
+          * If we try to move child to delete children into list before subclasses desruction,
+          * first child use count will be 2.
+          */
         {
-            /// If two threads remove ASTPtr concurrently,
-            /// it is possible that neither thead will see use_count == 1.
-            /// It is ok. Will need one more extra stack frame in this case.
-            if (node.use_count() == 1)
+            ASTPtr child_to_delete;
+            child_to_delete.swap(delete_list_head_reference);
```

Comment:
I did not understand why we have to swap here (and then traverse a list again).
Can't we just set `child_to_delete->next_to_delete_list_head` to `&next_node_to_delete`? And use itself only if next_node_to_delete is empty.

Reply:
Good catch. I did not change this place after I changed list append strategy, before I passed tail and added to tail, now as we replace list head with new list head and point it to previous head, we can safely do this.
We can write it like this:
```
ASTPtr child_to_delete;
child_to_delete.swap(delete_list_head_reference);

delete_list_head_reference = std::move(child_to_delete->next_to_delete);
child_to_delete->next_to_delete_list_head = &delete_list_head_reference;
```
Related to swap, I actually want to be more explicit about what is going on here. Updated comments.